src/index.ts

```diff
@@ -0,0 +1,518 @@
+   * @param translations  Nested JSON object, where object values are treated as additional context
+   *                      and string values are treated as translations. Other values are not valid.
+   */
+  public setTranslations(locale: string | string[], translations: Translations): void {
+    const locales = this.normalizeLocales(locale);
+    const clonedTranslations = this.cloneTranslations(translations);
+
+    locales.forEach((l) => {
+      this.translations[l] = clonedTranslations;
+    });
+
+    this.callSubscribers();
+  }
+
+  /**
+   * Subscribers are called when any of the internal state changes.
+   *
+   * @param handler Callback method for whenever a change occurs in the code.
+   * @returns       A method that when called, will unsubscribe the handler.
+   */
+  public subscribe(handler: Function): Unsubscribe {
+    this.subscribers.push(handler);
+
+    return () => {
+      const index: number = this.subscribers.indexOf(handler);
+      this.subscribers.splice(index, 1);
+    };
+  }
+
+  /**
+   * Find an appropriate string for a given key. This function will attempt to find the string
+   * using the primary locale, then will use the fallback if there is no match. If there are no
+   * matches, then the key will be returned.
+   *
+   * @param key   The key for looking up the string. Context should be delimited with a period.
+   * @param data  Contains data to be interpolated. `count` is a magic value for pluralization.
+   * @returns     Best string match for key given.
+   */
+  public t(key: string, data: TDataObject = {}, options: TOptions = {}): string {
+    // If we're not in production mode, be sure the people testing
+    if (process.env.NODE_ENV !== 'production') {
+      if (!key.match(VALID_KEY)) {
+        throw new Error(
+          `${key} is not a valid key format. ` +
+          `Valid keys are of the format ${VALID_KEY.toString()}`);
+      }
+    }
+
+    // If key mode is enabled, simply return the key.
+    if (this.getKeyMode()) {
+      return key;
+    }
+
+    const locales: string[] = [];
+
+    // If options has a locale, normalize it and use that instead of locale. Otherwise, check
+    // if locale is even set and add that to possible locales.
+    if (options.locale !== undefined) {
+      locales.push(this.normalizeLocale(options.locale));
+    } else if (this.locale !== undefined) {
+      locales.push(this.locale);
+    }
+
+    if (options.fallback !== undefined) {
+      locales.push(this.normalizeLocale(options.fallback));
+    } else {
+      locales.push(this.fallback);
+    }
+
+    // Loop through the locales and try to find an appropriate translation.
+    for (let i = 0; i < locales.length; i++) {
+      const locale = locales[i];
+      const contexts = key.split('.');
+
+      let translation: TranslationLookup =
+        contexts.reduce(this.translationReducer, this.translations[locale]);
```

Comment:
TIL about [Function Paramater Bivariance][]. This should have been a type error since the first argument of translation reducer does not accept the type of the expression `this.translations[locale]`.

[Function Paramater Bivariance]: https://www.typescriptlang.org/docs/handbook/type-compatibility.html#function-parameter-bivariance